drafts/25-xxx-macro-expansion.txt

```diff
@@ -0,0 +1,674 @@
+         the Fortran preprocessor (FPP)
+Date: 2025-XXX-XX
+References:
+      25-142r2 Formal specifications for the Fortran preprocessor (FPP)
+      25-114r2 Fortran preprocessor requirements
+      ISO/IEC 9899:2024 Programming languages -- C ("C 2023")
+            (working draft N3220)
+
+Background
+==========
+
+At meeting #235 in Feb 2025, J3 approved requirements for a cpp-like
+preprocessing for Fortran 202Y (paper 25-114r2).
+
+At meeting #236 in Jun 2025, J3 approved corresponding specifications
+covering the majority of preprocessing (paper 25-142r2). That document
+left one incomplete section, section 4 "Macro identification and
+expansion".
+
+This document provides the remaining specifications for preprocessing,
+specifying the details of macro identification and expansion. It should
+be read along with 25-142r2, which provides all the other
+specifications.
+
+
+1 General
+=========
+
+TODO: link to existing papers, explain this is a section of
+      already-passed document 25-142r2
+
+ge01. Macro identification and expansion is a process that occurs on
+      source fragment lines and comment lines when they are first
+      encountered during preprocessing, and may be recursively repeated
+      as specified in section 4.
+
+ge02. During expansion, an identifier that is the name of a defined
+      object-like macro (25-142r2 section 3.1) is replaced by the
+      replacement-list of that macro's definition.
+
+ge04. During expansion, an identifier that is the name of a defined
+      function-like macro (25-142r2 section 3.2) followed by a left
+      parenthesis '(' introduces an invocation of the function-like
+      macro. The arguments of the macro invocation are collected before
+      the macro is expanded (section 2).
+
+ge05. After expansion has replaced a macro invocation, the resulting
+      tokens are rescanned to find additional macro invocations for
+      expansion (section 4).
+
+
+
+2 Function-like macro invocation
+================================
+
+
+2.1 Function-like macro identification
+--------------------------------------
+
+2.1.1 Argument gathering
+------------------------
+
+ag01. During expansion when a function-like macro name is encountered
+      followed by a '(' as the next nonblank character, the processor
+      shall scan ahead, removing comments (section 2.1.4), to identify
+      the full list of macro arguments. The argument list begins after
+      the initial '(' and terminates at the matching closing parenthesis
+      ')'. The argument list shall terminate before the end of the
+      current file.
+
+ag03. Commas within the argument list separate individual arguments.
+      However, a comma that appears within a balanced set of parentheses
+      (), square brackets '[]', curly braces '{}', or array constructor
+      delimiters '(/ /)' does not act as an argument separator. All such
+      bracketing characters within an argument must be balanced and
+      properly nested.
```

Comment:
@bonachea we need to specify whether `(/` and `( /` are different for the purposes of determining nesting and balancing.